File: core/RemoveOutermostParentheses/RemoveOutermostParentheses.java
package com.dbc;

import java.util.Stack;

public class RemoveOutermostParentheses {
    public String removeOuterParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
        StringBuffer res = new StringBuffer();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') stack.add(i);
            else {
                int preIndex = stack.pop();
                if (stack.isEmpty()) res.append(s.substring(preIndex + 1, i));
            }
        }
        return res.toString();
    }
}

File: core/RemoveOutermostParentheses/RemoveOutermostParentheses.py
def removeOuterParentheses(s: str) -> str:
    res, stack, length = "", [], len(s)
    for index, ch in enumerate(s):
        if ch == "(":
            stack.append(index)
        else:
            pre_index = stack.pop()
            if not stack:
                res += s[pre_index + 1:index]
    return res





